### ISSUE TITLE
Components randomly get removed from prefabs in between sessions

### ISSUE THREAD

**Issue by yuberee** [Facepunch/sbox-issues#4273]
### Describe the bug

Whenever you start a new sessions there's a chance components get removed from your prefabs.
When opening a prefab the components you know were there do not appear, but looking at changes through a version control software reveals that nothing changed. Only after saving the 

### To Reproduce

This is seemingly random and has no reproduction, but I managed to catch it in action in a separate branch, and hopefully it can get replicated.

Here's the branch to clone showcased in the video below: https://github.com/Small-Fish-Dev/Snot-Farm/tree/Bug-Repro

Basically open any of the prefabs and save with CTRL+S

### Expected behavior

.

### Media/Files

https://github.com/Facepunch/sbox-issues/assets/59583743/02f75f3b-4b24-4ec1-b56c-4db51a95ec14

### Additional context

Here's the log of the session where this has happened multiple times: [Log.log](https://github.com/Facepunch/sbox-issues/files/13732532/Log.log)

**Comment by yuberee** [comment]
![image](https://github.com/Facepunch/sbox-issues/assets/59583743/8bbc0825-38e5-40b9-9d2c-ca7f55047455)
Well I reloaded the project and it doesn't happen anymore...
Here's the log where it doesn't happen just for reference:
[Log.log](https://github.com/Facepunch/sbox-issues/files/13732597/Log.log)

**Comment by NotGaylien** [comment]
https://github.com/Facepunch/sbox-issues/issues/4268
https://github.com/Facepunch/sbox-issues/issues/4178

think these are related.
stuff not being loaded properly when opening and or switching projects

**Comment by garrynewman** [comment]
I can imagine how switching projects would cause this. Is it likely that this is happening when switching projects?

**Comment by yuberee** [comment]
I can't say for the others but lately I've been working on the same project and it has happened twice.

**Comment by garrynewman** [comment]
I've pushed some changes that I suspect will fix this

**Comment by garrynewman** [comment]
So this was apparently still happening. There was an issue where it let you load a game project when the game project wasn't compiling. I've plugged those gaps - so putting this back in review now to see if it still happens.



![Image](https://github.com/Facepunch/sbox-issues/assets/928843/139c0c27-96eb-46d4-9f09-8915491372fd)

**Comment by garrynewman** [comment]
This seems to have stopped being a problem now, right?

**Comment by yuberee** [comment]
True! It was happening basically everytime and now I haven't seen it once.
I'll close it, if it happens again in the future it's probably going to be something different.

**Comment by yuberee** [comment]
It seems to have come back.

```
TypeLibrary could not find GridAStar.Grid
   at Sandbox.Component.Serialize(SerializeOptions options)
   at Sandbox.GameObject.Serialize(SerializeOptions options)
   at Sandbox.Scene.<>c.<Save>b__77_0(GameObject x)
   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at Sandbox.Scene.Save()
   at Sandbox.SceneExtensions.Save(Scene scene, Boolean saveAs)
   at Editor.SceneEditorMenus.SaveScene() in C:\Program Files (x86)\Steam\steamapps\common\sbox\addons\tools\code\Scene\SceneEditorMenus.cs:line 15
   at InvokeStub_SceneEditorMenus.SaveScene(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Sandbox.MethodDescription.Invoke(Object targetObject, Object[] parameters)
   at Editor.MenuAttribute.<>c__DisplayClass29_0.<Register>b__0()
   at Editor.Option.OnTriggered()
   at Editor.Option.InternalTriggered()
   at Managed.SourceTools.Exports.Editor_Option_InternalTriggered(UInt32 self)
   at Sandbox.EngineLoop.RunFrame(CMaterialSystem2AppSystemDict appDict, Boolean& wantsQuit)
   at Sandbox.SourceEngineApp.RunFrame() in C:\build\_work\sbox\sbox\engine\Launcher\SourceEngineApp.cs:line 68
   at Sandbox.SourceEngineApp.RunLoop() in C:\build\_work\sbox\sbox\engine\Launcher\SourceEngineApp.cs:line 57
   at Sandbox.Program.LaunchGame() in C:\build\_work\sbox\sbox\engine\Launcher\Launcher.cs:line 91
   at Sandbox.Program.Main() in C:\build\_work\sbox\sbox\engine\Launcher\Launcher.cs:line 57
```

![image](https://github.com/Facepunch/sbox-issues/assets/59583743/877cc166-eb4d-4aed-9160-c292869c3322)

This is the repo I'm working on: https://github.com/Small-Fish-Dev/Grid-and-Astar-NPC/tree/scene
Restarting doesn't fix it, has been happening since the first time I created this component.
I'm thinking it's because it's a library and I'm working on it through a game project since content mode is broken.

**Comment by yuberee** [comment]
This has suddenly fixed itself moments after opening the issue 😅

**Comment by garrynewman** [comment]
Working via a library isn't really supported right now

**Comment by MrBrax** [comment]
still happens to me, gotten worse recently. working on a game project and not a library.
````
TypeLibrary couldn't find Component type Money
   at Sandbox.GameObject.Deserialize(JsonObject node)
   at Sandbox.Scene.Load(SceneLoadOptions options)
   at Sandbox.Scene.Load(GameResource resource)
   at Editor.EditorScene.LoadFromScene(SceneFile resource)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Editor.EditorScene.LoadFromScene(SceneFile resource)
   at Editor.EditorScene.GameStarted(Package package)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Editor.EditorScene.GameStarted(Package package)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Sandbox.Internal.EventSystem.<>c__DisplayClass11_0.<BuildDelegate>b__0(Object o, Object[] p)
   at Sandbox.Internal.EventSystem.Run(String v, Object[] list)
   at Editor.EditorEvent.Run[T](String name, T arg0)
   at Sandbox.ToolsDll.OnGameStarted(Package package)
   at Sandbox.GameMenu.GameMenuDll.LoadGamePackageAsync(String ident, GameLoadingFlags flags, CancellationToken ct)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at Sandbox.Tasks.ExpirableSynchronizationContext.ProcessQueue()
   at Sandbox.Tasks.ExpirableSynchronizationContext.Wait(IntPtr[] waitHandles, Boolean waitAll, Int32 millisecondsTimeout)
   at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
   at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
   at Sandbox.AccessControl.TestModule(ModuleDefinition module)
   at Sandbox.AccessControl.InitTouches(Byte[] dll)
   at Sandbox.AccessControl.VerifyAssembly(Stream dll, TrustedBinaryStream& outStream, Boolean addToWhitelist)
   at Sandbox.PackageLoader.TestAccessControl(Stream dll, TrustedBinaryStream& trustedDll)
   at Sandbox.PackageLoader.LoadAssemblyFromPackage(ActivePackage ap, String filename)
   at Sandbox.PackageLoader.LoadPendingChanges()
   at Sandbox.PackageLoader.Tick()
   at Sandbox.GameMenu.GameMenuDll.FinishLoadingAssemblies()
   at Sandbox.GameMenu.GameMenuDll.Tick()
   at Sandbox.EngineLoop.FrameStart()
   at Sandbox.EngineLoop.RunFrame(CMaterialSystem2AppSystemDict appDict, Boolean& wantsQuit)
   at Sandbox.SourceEngineApp.RunFrame() in C:\build\_work\sbox\sbox\engine\Launcher\SourceEngineApp.cs:line 68
   at Sandbox.SourceEngineApp.RunLoop() in C:\build\_work\sbox\sbox\engine\Launcher\SourceEngineApp.cs:line 57
   at Sandbox.Program.LaunchGame() in C:\build\_work\sbox\sbox\engine\Launcher\Launcher.cs:line 91
   at Sandbox.Program.Main() in C:\build\_work\sbox\sbox\engine\Launcher\Launcher.cs:line 57
````

**Comment by garrynewman** [comment]
While this might happen now, it should no longer be destructive

**Comment by MrBrax** [comment]
*might* is an understatement as it completely breaks the entire project with seemingly no way to solve it now, but the data being kept is really nice

**Comment by garrynewman** [comment]
If you find a repro please make a new issue.

**Comment by MrBrax** [comment]
ok, i thought it was a known issue

**Comment by anthonysharpy** [comment]
I had this just now, here's how I fixed it, maybe it will give some clues

- Opened up project, every time I pressed play it was saying failed to find component type blah blah, and if I tried to search for my components by pressing the add component button, none of my custom components were showing up
- I opened up VS while s&box was open in the background with my scene, s&box then somehow started compiling by itself
- After it was done compiling s&box could find my components again and it started working

**Comment by yuberee** [comment]
Seems like the other possible variation has been fixed as well. Closing this for now.
